### PR TITLE
Optimize discriminator predicates into IN expressions in more cases

### PR DESCRIPTION
**Pull request by tuespetre**
This PR isolates one of the optimizations I made in #7543 into a smaller commit. It will shave two files off of that PR. This PR partially address #5230.

### CODE REVIEW

File: src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/Internal/CompositePredicateExpressionVisitor.cs [hunk 1]
```diff
@@ -28,29 +28,33 @@ public CompositePredicateExpressionVisitor(bool useRelationalNulls)
         ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
         ///     directly from your code. This API may change or be removed in future releases.
         /// </summary>
-        public override Expression Visit(Expression expression)
+        protected override Expression VisitExtension(Expression expression)
         {
-            if (expression != null)
+            var selectExpression = expression as SelectExpression;
+
+            if (selectExpression?.Predicate != null)
```
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/Internal/CompositePredicateExpressionVisitor.cs]: This change in a way scoping down the optimizations to only predicate of select expression (applied recursively)

**Comment by maumar** [comment]
This PR actually addresses https://github.com/aspnet/EntityFramework/issues/7548 - can you add test case for it? (also reflect this in the commit message)

File: src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/Internal/CompositePredicateExpressionVisitor.cs (hunk 1 above)
Comment (maumar) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/Internal/CompositePredicateExpressionVisitor.cs]: +1 to what @smitpatel said. Please remove this for now, we need to look into the CompositePredicateExpressionVisitor in general, rename it and see where the optimizations can be applied to other parts of the SelectExpression. This can be done in a separate PR tho.

File: src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/Internal/IncludeExpressionVisitor.cs [hunk 2]
```diff
@@ -277,9 +277,7 @@ var materializer
 
                     if (selectExpression.Predicate != oldPredicate)
                     {
-                        selectExpression.Predicate
-                            = compositePredicateExpressionVisitor
-                                .Visit(selectExpression.Predicate);
+                        compositePredicateExpressionVisitor.Visit(selectExpression);
```
Comment (maumar) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/Internal/IncludeExpressionVisitor.cs]: please revert these changes for now

File: src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs [hunk 3]
```diff
@@ -390,8 +390,7 @@ var typeIsExpressionTranslatingVisitor
 
             foreach (var selectExpression in QueriesBySource.Values)
             {
-                selectExpression.Predicate
-                    = compositePredicateVisitor.Visit(selectExpression.Predicate);
+                compositePredicateVisitor.Visit(selectExpression);
```
Comment (maumar) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs]: please revert these changes for now

**Comment by tuespetre** [comment]
Would you mind if I submit a new PR for the other test case and its solution? The changes you requested here remove the optimizations seen in the test SQL output anyways, it's not really the same PR anymore.

**Comment by tuespetre** [comment]
*not all of the optimizations I suppose, but any of the optimizations occurring in subqueries. I will make the changes here then.

**Comment by maumar** [comment]
@tuespetre after giving is some more thought, perhaps your current solution is fine. Before we were applying it only to top level predicates, so we won't regress any optimizations, and since in your solution optimizations are still only applied to the predicates (just recursively), the name of the visitor is still adequate.

I will file an issue to track cleanup of this for later - i.e. to see if we can apply some of the optimizations in more general fashion, e.g. to projections.

Please add the extra test case in this PR and it should be good to do, sorry for randomizing you ;)

**Comment by smitpatel** [comment]
`var maumar = new Random();` :trollface: 

@tuespetre - In general, there is no harm in creating new PR, but when tackling certain issue, we tend to use the same PR even if implementation changes. That is done to preserve the PR feedback comment. In case of totally change of implementation which would make the feedback on PR totally irrelevant, you should file new PR.

**Comment by tuespetre** [comment]
:up::date:

**Comment by smitpatel** [comment]
filed #7612
